File: desktop/app/orchestrator/kpi.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from uuid import uuid4

from app.chat.test_user import ILCHENKO_USER_ID, is_ilchenko_user
from app.orchestrator.models import COMPLETED, ERROR, MEETING_ID, READY, REVISION_ID, WAITING_HUMAN, ProcessInstance

_GREEN = "#08745F"
_YELLOW = "#C9A227"
_RED = "#C0392B"


@dataclass(frozen=True)
class KpiDefinition:
    number: int
    name: str
    target: float
    weight: float
    kind: str


@dataclass(frozen=True)
class KpiRow:
    number: int
    name: str
    target: float
    weight: float
    fact: float | None
    color: str


ILCHENKO_KPI: tuple[KpiDefinition, ...] = (
    KpiDefinition(1, "Своевременность пакета к заседаниям (СД + РК)", 95, 25, "package_on_time"),
    KpiDefinition(2, "Своевременность протоколов (СД + РК)", 95, 25, "protocol_on_time"),
    KpiDefinition(3, "Реестр и контроль исполнения поручений (СД + РК)", 95, 25, "instructions"),
    KpiDefinition(4, "Качество протокола и материалов (без возвратов по замечаниям)", 98, 25, "quality"),
)


def has_position_kpi(user_id: str = "", fio: str = "") -> bool:
    return user_id == ILCHENKO_USER_ID or is_ilchenko_user(user_id, fio)


def _ratio(ok: int, total: int) -> float | None:
    if total <= 0:
        return None
    return 100.0 * ok / total


def _closed(instances: list[ProcessInstance]) -> list[ProcessInstance]:
    return [item for item in instances if item.status == COMPLETED]


def _has_return(instance: ProcessInstance) -> bool:
    return any(str(event.get("type") or "") == "returned" for event in instance.events)


def _fact(kind: str, instances: list[ProcessInstance]) -> float | None:
    if kind in {"package_on_time", "protocol_on_time"}:
        done = [item for item in instances if item.status in {COMPLETED, ERROR}]
        return _ratio(sum(1 for item in done if item.status == COMPLETED), len(done))
    if kind == "instructions":
        started = [item for item in instances if item.status != READY]
        return _ratio(len(_closed(instances)), len(started))
    if kind == "quality":
        closed = _closed(instances)
        return _ratio(sum(1 for item in closed if not _has_return(item)), len(closed))
    return None


def _color(fact: float | None, target: float) -> str:
    if fact is None:
        return "#6B7773"
    if fact + 1e-9 >= target:
        return _GREEN
    if fact + 1e-9 >= target - 10:
        return _YELLOW
    return _RED


def score_rows(instances: list[ProcessInstance], definitions: tuple[KpiDefinition, ...] = ILCHENKO_KPI) -> list[KpiRow]:
    rows: list[KpiRow] = []
    for item in definitions:
        fact = _fact(item.kind, instances)
        rows.append(
            KpiRow(
                number=item.number,
                name=item.name,
                target=item.target,
                weight=item.weight,
                fact=fact,
                color=_color(fact, item.target),
            )
        )
    return rows


def weighted_score(rows: list[KpiRow]) -> float | None:
    total_weight = 0.0
    acc = 0.0
    for row in rows:
        if row.fact is None:
            continue
        acc += row.fact * row.weight
        total_weight += row.weight
    if total_weight <= 0:
        return None
    return acc / total_weight


def format_percent(value: float | None) -> str:
    if value is None:
        return "нет данных"
    return f"{value:.1f}%".replace(".0%", "%")


def seed_ilchenko_instances() -> list[ProcessInstance]:
    now = datetime.now(timezone.utc).isoformat()
    rows: list[ProcessInstance] = []

    def add(definition_id: str, status: str, events: list[dict], waiting: int = 0) -> None:
        rows.append(
            ProcessInstance(
                id=str(uuid4()),
                definition_id=definition_id,
                status=status,
                waiting=waiting,
                updated_at=now,
                events=events,
            )
        )

    for _ in range(8):
        add(MEETING_ID, COMPLETED, [{"type": "approved", "status": COMPLETED, "at": now}])
    for _ in range(7):
        add(REVISION_ID, COMPLETED, [{"type": "approved", "status": COMPLETED, "at": now}])
    add(
        MEETING_ID,
        COMPLETED,
        [
            {"type": "returned", "status": "ACTIVE", "at": now},
            {"type": "approved", "status": COMPLETED, "at": now},
        ],
    )
    add(REVISION_ID, WAITING_HUMAN, [{"type": "seed", "status": WAITING_HUMAN, "at": now}], waiting=1)
    add(MEETING_ID, READY, [{"type": "seed", "status": READY, "at": now}])
    return rows

File: desktop-electron/src/renderer/src/orchestrator/kpi.ts
import {
  COMPLETED,
  ERROR,
  MEETING_ID,
  READY,
  REVISION_ID,
  WAITING_HUMAN,
  type ProcessInstance
} from './models'

export const ILCHENKO_USER_IDS = new Set([
  'A2DCC949FEDEC70D40318ABA83C618F4',
  'E11C4E11K00000000000000000000001'
])

export interface KpiRow {
  number: number
  name: string
  target: number
  weight: number
  fact: number | null
  color: string
}

const KPI = [
  { number: 1, name: 'Своевременность пакета к заседаниям (СД + РК)', target: 95, weight: 25, kind: 'package_on_time' },
  { number: 2, name: 'Своевременность протоколов (СД + РК)', target: 95, weight: 25, kind: 'protocol_on_time' },
  { number: 3, name: 'Реестр и контроль исполнения поручений (СД + РК)', target: 95, weight: 25, kind: 'instructions' },
  { number: 4, name: 'Качество протокола и материалов (без возвратов по замечаниям)', target: 98, weight: 25, kind: 'quality' }
] as const

export function hasPositionKpi(userId = '', fio = ''): boolean {
  if (ILCHENKO_USER_IDS.has((userId || '').trim())) return true
  return (fio || '').toLowerCase().includes('ильченко')
}

function ratio(ok: number, total: number): number | null {
  if (total <= 0) return null
  return (100 * ok) / total
}

function closed(instances: ProcessInstance[]): ProcessInstance[] {
  return instances.filter((item) => item.status === COMPLETED)
}

function hasReturn(instance: ProcessInstance): boolean {
  return instance.events.some((event) => String(event.type || '') === 'returned')
}

function fact(kind: string, instances: ProcessInstance[]): number | null {
  if (kind === 'package_on_time' || kind === 'protocol_on_time') {
    const done = instances.filter((item) => item.status === COMPLETED || item.status === ERROR)
    return ratio(done.filter((item) => item.status === COMPLETED).length, done.length)
  }
  if (kind === 'instructions') {
    const started = instances.filter((item) => item.status !== READY)
    return ratio(closed(instances).length, started.length)
  }
  if (kind === 'quality') {
    const done = closed(instances)
    return ratio(done.filter((item) => !hasReturn(item)).length, done.length)
  }
  return null
}

function color(value: number | null, target: number): string {
  if (value == null) return '#6B7773'
  if (value + 1e-9 >= target) return '#08745F'
  if (value + 1e-9 >= target - 10) return '#C9A227'
  return '#C0392B'
}

export function scoreRows(instances: ProcessInstance[]): KpiRow[] {
  return KPI.map((item) => {
    const value = fact(item.kind, instances)
    return {
      number: item.number,
      name: item.name,
      target: item.target,
      weight: item.weight,
      fact: value,
      color: color(value, item.target)
    }
  })
}

export function weightedScore(rows: KpiRow[]): number | null {
  let total = 0
  let acc = 0
  for (const row of rows) {
    if (row.fact == null) continue
    acc += row.fact * row.weight
    total += row.weight
  }
  return total > 0 ? acc / total : null
}

export function formatPercent(value: number | null): string {
  if (value == null) return 'нет данных'
  return `${value.toFixed(1)}%`.replace('.0%', '%')
}

export function seedIlchenkoInstances(): ProcessInstance[] {
  const now = new Date().toISOString()
  const rows: ProcessInstance[] = []
  const add = (
    definitionId: string,
    status: string,
    events: Array<Record<string, string>>,
    waiting = 0
  ): void => {
    rows.push({
      id: crypto.randomUUID(),
      definition_id: definitionId,
      status,
      waiting,
      updated_at: now,
      events
    })
  }
  for (let i = 0; i < 8; i += 1) {
    add(MEETING_ID, COMPLETED, [{ type: 'approved', status: COMPLETED, at: now }])
  }
  for (let i = 0; i < 7; i += 1) {
    add(REVISION_ID, COMPLETED, [{ type: 'approved', status: COMPLETED, at: now }])
  }
  add(MEETING_ID, COMPLETED, [
    { type: 'returned', status: 'ACTIVE', at: now },
    { type: 'approved', status: COMPLETED, at: now }
  ])
  add(REVISION_ID, WAITING_HUMAN, [{ type: 'seed', status: WAITING_HUMAN, at: now }], 1)
  add(MEETING_ID, READY, [{ type: 'seed', status: READY, at: now }])
  return rows
}
